docs(guide): adding guide notebooks and supporting them with fixes (#9)
Deltakit/deltakit: docs/guide/adding_noise.md

---
jupytext:
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.17.2
kernelspec:
  name: python3
  display_name: Python 3 (ipykernel)
  language: python
---

# Adding Noise

## Motivation

> Noise is an essential component of QEC workflows.
> The whole idea of QEC only makes sense if qubits are not perfect.
> Decoders try to make logical qubits robust against errors,
> but in order for them to do a good job we have to inform them which errors they should expect.

Reviewer: ```suggestion
The goal of Quantum Error Correction is to correct errors, which are caused by different noise channels in the system. In order for a decoder to more accurately approximate errors in the system based on a set of error syndromes, it's important to pass the characteristics of these noise channels.
```

Author: May you consider my proposal of your edited text.
```suggestion
The goal of Quantum Error Correction is to correct errors, which are caused by different noise mechanisms.
A noise channel is a way to describe a probabilistic change of a quantum state.
In order for a decoder to more accurately correct errors,
it's important to pass a model of the noise, described using noise channels.
```

```

> While exploring QEC codes or designing a new decoder,
> you should be able to validate it on real data.
> In the best case, you may run an experiment on a real QPU.
> But even if you have access to a QPU, it typically spends
> more time calibrating than doing computation.
> That is why you are interested in realistic QEC simulation.

Reviewer: ```suggestion
Whether you are exploring error-correcting codes or designing a new decoder,
it's important to have realistic data to validate them.
You can obtain those data from a real QPU, however, this might be expensive or time consuming, due to requiring calibration. This is where realistic QEC simulation comes in.
```

> The physics of noise in quantum systems is complicated and
> differs between qubit architectures.
> Superconducting systems don't typically lose their qubits,
> while neutral atom ones can.
> Cross-talk, on the other hand, is specific to
> superconducting QPUs.
> Shuttling noise is something you would observe in ion trap devices,
> but cannot even define in superconducting systems,
> and so on.

Reviewer: ```suggestion
The physics of noise in quantum systems is complicated and
differs between qubit architectures.
For example, superconducting systems don't typically lose their qubits,
while neutral atom ones can.
Cross-talk, on the other hand, is specific to
superconducting QPUs.
Shuttling noise is something you would observe in ion trap devices,
but cannot even be defined in superconducting systems,
and so on.
```

Each noise source has its own theory, and you may try to model it quite accurately.
The problem comes with the experiment size.
As you explore probabilistic processes, you want your evidence to be statistically reliable.
But due to the exponential improvement in logical error-probability when scaling an error-correction code,
typical experiment repetition numbers (also known as number of shots) may exceed $10^8$.
Doing accurate simulation with this number of repetitions is practically impossible,
so you may have to sacrifice model accuracy for simulation speed.

## Practical solution

> The `Stim` library offers tools for both quick simulation (thanks to its limitation to Clifford gates)

Reviewer: ```suggestion
The [Stim](https://github.com/quantumlib/Stim) library offers tools for both quick simulation (thanks to its limitation to Clifford gates)
```

and a reasonably rich palette of error mechanisms,
which may be used to approximate different types of noise.
Deltakit benefits from both Stim's simulation and error implementation approaches.
You may read about these building blocks in detail in the
[Noise Channels section](https://github.com/quantumlib/Stim/blob/main/doc/gates.md#noise-channels)
of the Stim documentation.

With these tools in your hands, you may implement different **noise models**.
A noise model is a way to define realistic errors across all qubits and gates
using a small number of controllable parameters.

## `NoiseParameters` class
The {class}`NoiseParameters <deltakit.explorer.qpu.NoiseParameters>` base class
allows you to define arbitrary functions which insert
noise events into the circuit when the necessary conditions are met.
You may define your realistic noise model by specifying what should happen
each time a gate or measurement is executed, or when a qubit remains idle.
Please refer to the documentation for this class.

All other noise models in Deltakit inherit from this class.
It defines a uniform way of adding Stim noise channels to the circuit.

## `ToyNoise`

> The {class}`ToyNoise <deltakit.explorer.qpu.ToyNoise>` noise model
> has only two parameters.

Reviewer: ```suggestion
The {class}`ToyNoise <deltakit.explorer.qpu.ToyNoise>` is a simple toy model for noise that
has only two parameters.
```
